File: im-backend/src/main/java/com/im/backend/modules/local_life/entity/MerchantReview.java
package com.im.backend.modules.local_life.entity;

import com.baomidou.mybatisplus.annotation.*;
import com.im.backend.common.BaseEntity;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

/**
 * 商户评价实体类
 * 支持多维度评价体系：星级评分 + 细评维度
 */
@Data
@EqualsAndHashCode(callSuper = true)
@TableName("merchant_review")
public class MerchantReview extends BaseEntity {

    /** 评价ID */
    @TableId(type = IdType.ASSIGN_ID)
    private Long id;

    /** 商户ID */
    private Long merchantId;

    /** POI兴趣点ID */
    private Long poiId;

    /** 用户ID */
    private Long userId;

    /** 订单ID（可选，关联消费订单） */
    private Long orderId;

    /** 综合星级评分（1-5分，支持半星） */
    private BigDecimal overallRating;

    /** 口味评分（1-5分，餐饮类商户） */
    private BigDecimal tasteRating;

    /** 环境评分（1-5分） */
    private BigDecimal environmentRating;

    /** 服务评分（1-5分） */
    private BigDecimal serviceRating;

    /** 性价比评分（1-5分） */
    private BigDecimal valueRating;

    /** 评价内容文本 */
    private String content;

    /** 评价图片URLs（JSON数组存储） */
    private String images;

    /** 评价视频URL */
    private String videoUrl;

    /** 视频封面图 */
    private String videoCover;

    /** 视频时长（秒） */
    private Integer videoDuration;

    /** 消费金额 */
    private BigDecimal consumptionAmount;

    /** 人均消费 */
    private BigDecimal perCapitaCost;

    /** 是否匿名评价 */
    private Boolean anonymous;

    /** 评价类型：1-图文评价 2-视频评价 */
    private Integer reviewType;

    /** 评价来源：1-APP 2-小程序 3-H5 */
    private Integer source;

    /** 点赞数 */
    private Integer likeCount;

    /** 回复数 */
    private Integer replyCount;

    /** 浏览数 */
    private Integer viewCount;

    /** 是否推荐（平台优质评价标记） */
    private Boolean recommended;

    /** 推荐权重（用于排序） */
    private Integer recommendWeight;

    /** 评价状态：0-待审核 1-已通过 2-已拒绝 3-已删除 */
    private Integer status;

    /** 审核备注 */
    private String auditRemark;

    /** 审核时间 */
    private LocalDateTime auditTime;

    /** 审核人ID */
    private Long auditBy;

    /** 是否置顶 */
    private Boolean pinned;

    /** 置顶排序 */
    private Integer pinOrder;

    /** 用餐时间（餐饮类） */
    private LocalDateTime diningTime;

    /** 用餐人数 */
    private Integer diningPeople;

    /** 标签（JSON数组：口味赞、环境好等） */
    private String tags;

    /** 是否体验过 */
    private Boolean experienced;

    /** 体验方式：1-到店消费 2-外卖 3-预约到店 */
    private Integer experienceType;

    /** 地理位置-经度 */
    private BigDecimal longitude;

    /** 地理位置-纬度 */
    private BigDecimal latitude;

    /** 地理位置名称 */
    private String locationName;

    /** 是否已同步到ES */
    private Boolean syncedToEs;

    /** 同步ES时间 */
    private LocalDateTime esSyncTime;

    /** 创建时间 */
    @TableField(fill = FieldFill.INSERT)
    private LocalDateTime createTime;

    /** 更新时间 */
    @TableField(fill = FieldFill.INSERT_UPDATE)
    private LocalDateTime updateTime;

    /** 逻辑删除 */
    @TableLogic
    @TableField(fill = FieldFill.INSERT)
    private Boolean deleted;

    // ========== 非持久化字段 ==========

    /** 用户昵称 */
    @TableField(exist = false)
    private String userNickname;

    /** 用户头像 */
    @TableField(exist = false)
    private String userAvatar;

    /** 用户等级 */
    @TableField(exist = false)
    private Integer userLevel;

    /** 商户名称 */
    @TableField(exist = false)
    private String merchantName;

    /** 商户Logo */
    @TableField(exist = false)
    private String merchantLogo;

    /** 是否已点赞 */
    @TableField(exist = false)
    private Boolean hasLiked;

    /** 图片列表（解析后的） */
    @TableField(exist = false)
    private List<String> imageList;

    /** 标签列表（解析后的） */
    @TableField(exist = false)
    private List<String> tagList;

    /** 回复列表 */
    @TableField(exist = false)
    private List<MerchantReviewReply> replies;

    /**
     * 计算平均分
     */
    public BigDecimal calculateAverageRating() {
        if (tasteRating == null && environmentRating == null && 
            serviceRating == null && valueRating == null) {
            return overallRating;
        }
        
        BigDecimal sum = BigDecimal.ZERO;
        int count = 0;
        
        if (tasteRating != null) {
            sum = sum.add(tasteRating);
            count++;
        }
        if (environmentRating != null) {
            sum = sum.add(environmentRating);
            count++;
        }
        if (serviceRating != null) {
            sum = sum.add(serviceRating);
            count++;
        }
        if (valueRating != null) {
            sum = sum.add(valueRating);
            count++;
        }
        
        if (count == 0) {
            return overallRating;
        }
        
        return sum.divide(new BigDecimal(count), 1, BigDecimal.ROUND_HALF_UP);
    }

    /**
     * 获取星级文本描述
     */
    public String getRatingText() {
        if (overallRating == null) return "暂无评分";
        double rating = overallRating.doubleValue();
        if (rating >= 4.5) return "极佳";
        if (rating >= 4.0) return "非常好";
        if (rating >= 3.5) return "不错";
        if (rating >= 3.0) return "一般";
        if (rating >= 2.0) return "较差";
        return "很差";
    }

    /**
     * 是否为优质评价（字数+图片/视频）
     */
    public boolean isHighQuality() {
        boolean hasEnoughText = content != null && content.length() >= 50;
        boolean hasMedia = (images != null && !images.isEmpty()) || 
                          (videoUrl != null && !videoUrl.isEmpty());
        return hasEnoughText && hasMedia;
    }

    /**
     * 是否为视频评价
     */
    public boolean isVideoReview() {
        return videoUrl != null && !videoUrl.isEmpty();
    }
}

File: im-mobile/lib/models/merchant_review.dart
/// 商户评价模型
class MerchantReview {
  final int id;
  final int merchantId;
  final int? poiId;
  final int userId;
  final int? orderId;
  final double overallRating;
  final double? tasteRating;
  final double? environmentRating;
  final double? serviceRating;
  final double? valueRating;
  final String content;
  final List<String>? images;
  final String? videoUrl;
  final String? videoCover;
  final int? videoDuration;
  final double? consumptionAmount;
  final double? perCapitaCost;
  final bool anonymous;
  final int reviewType; // 1-图文 2-视频
  final int source;
  final int likeCount;
  final int replyCount;
  final int viewCount;
  final bool recommended;
  final DateTime? diningTime;
  final int? diningPeople;
  final List<String>? tags;
  final bool experienced;
  final int? experienceType;
  final DateTime createTime;
  
  // 用户信息
  final String? userNickname;
  final String? userAvatar;
  final int? userLevel;
  
  // 商户信息
  final String? merchantName;
  final String? merchantLogo;
  
  // 交互状态
  final bool? hasLiked;
  final List<MerchantReviewReply>? replies;
  final String? ratingText;
  final bool? highQuality;

  MerchantReview({
    required this.id,
    required this.merchantId,
    this.poiId,
    required this.userId,
    this.orderId,
    required this.overallRating,
    this.tasteRating,
    this.environmentRating,
    this.serviceRating,
    this.valueRating,
    required this.content,
    this.images,
    this.videoUrl,
    this.videoCover,
    this.videoDuration,
    this.consumptionAmount,
    this.perCapitaCost,
    this.anonymous = false,
    this.reviewType = 1,
    this.source = 1,
    this.likeCount = 0,
    this.replyCount = 0,
    this.viewCount = 0,
    this.recommended = false,
    this.diningTime,
    this.diningPeople,
    this.tags,
    this.experienced = false,
    this.experienceType,
    required this.createTime,
    this.userNickname,
    this.userAvatar,
    this.userLevel,
    this.merchantName,
    this.merchantLogo,
    this.hasLiked,
    this.replies,
    this.ratingText,
    this.highQuality,
  });

  factory MerchantReview.fromJson(Map<String, dynamic> json) {
    return MerchantReview(
      id: json['id'] as int,
      merchantId: json['merchantId'] as int,
      poiId: json['poiId'] as int?,
      userId: json['userId'] as int,
      orderId: json['orderId'] as int?,
      overallRating: (json['overallRating'] as num).toDouble(),
      tasteRating: json['tasteRating'] != null ? (json['tasteRating'] as num).toDouble() : null,
      environmentRating: json['environmentRating'] != null ? (json['environmentRating'] as num).toDouble() : null,
      serviceRating: json['serviceRating'] != null ? (json['serviceRating'] as num).toDouble() : null,
      valueRating: json['valueRating'] != null ? (json['valueRating'] as num).toDouble() : null,
      content: json['content'] as String,
      images: json['images'] != null ? List<String>.from(json['images']) : null,
      videoUrl: json['videoUrl'] as String?,
      videoCover: json['videoCover'] as String?,
      videoDuration: json['videoDuration'] as int?,
      consumptionAmount: json['consumptionAmount'] != null ? (json['consumptionAmount'] as num).toDouble() : null,
      perCapitaCost: json['perCapitaCost'] != null ? (json['perCapitaCost'] as num).toDouble() : null,
      anonymous: json['anonymous'] as bool? ?? false,
      reviewType: json['reviewType'] as int? ?? 1,
      source: json['source'] as int? ?? 1,
      likeCount: json['likeCount'] as int? ?? 0,
      replyCount: json['replyCount'] as int? ?? 0,
      viewCount: json['viewCount'] as int? ?? 0,
      recommended: json['recommended'] as bool? ?? false,
      diningTime: json['diningTime'] != null ? DateTime.parse(json['diningTime']) : null,
      diningPeople: json['diningPeople'] as int?,
      tags: json['tags'] != null ? List<String>.from(json['tags']) : null,
      experienced: json['experienced'] as bool? ?? false,
      experienceType: json['experienceType'] as int?,
      createTime: DateTime.parse(json['createTime']),
      userNickname: json['userNickname'] as String?,
      userAvatar: json['userAvatar'] as String?,
      userLevel: json['userLevel'] as int?,
      merchantName: json['merchantName'] as String?,
      merchantLogo: json['merchantLogo'] as String?,
      hasLiked: json['hasLiked'] as bool?,
      replies: json['replies'] != null 
          ? (json['replies'] as List).map((e) => MerchantReviewReply.fromJson(e)).toList()
          : null,
      ratingText: json['ratingText'] as String?,
      highQuality: json['highQuality'] as bool?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'merchantId': merchantId,
      'poiId': poiId,
      'userId': userId,
      'orderId': orderId,
      'overallRating': overallRating,
      'tasteRating': tasteRating,
      'environmentRating': environmentRating,
      'serviceRating': serviceRating,
      'valueRating': valueRating,
      'content': content,
      'images': images,
      'videoUrl': videoUrl,
      'videoCover': videoCover,
      'videoDuration': videoDuration,
      'consumptionAmount': consumptionAmount,
      'perCapitaCost': perCapitaCost,
      'anonymous': anonymous,
      'reviewType': reviewType,
      'source': source,
      'likeCount': likeCount,
      'replyCount': replyCount,
      'viewCount': viewCount,
      'recommended': recommended,
      'diningTime': diningTime?.toIso8601String(),
      'diningPeople': diningPeople,
      'tags': tags,
      'experienced': experienced,
      'experienceType': experienceType,
      'createTime': createTime.toIso8601String(),
    };
  }

  /// 计算平均分
  double calculateAverageRating() {
    final ratings = [tasteRating, environmentRating, serviceRating, valueRating]
        .where((r) => r != null)
        .toList();
    if (ratings.isEmpty) return overallRating;
    return ratings.reduce((a, b) => a! + b!)! / ratings.length;
  }

  /// 获取星级颜色
  String get ratingColor {
    if (overallRating >= 4.5) return '#FF6B6B'; // 极佳-红色
    if (overallRating >= 4.0) return '#FFA500'; // 很好-橙色
    if (overallRating >= 3.5) return '#FFD700'; // 不错-金色
    if (overallRating >= 3.0) return '#90EE90'; // 一般-浅绿
    return '#808080'; // 较差-灰色
  }

  /// 是否为视频评价
  bool get isVideoReview => videoUrl != null && videoUrl!.isNotEmpty;

  /// 是否有图片
  bool get hasImages => images != null && images!.isNotEmpty;

  /// 获取显示标题
  String get displayTitle {
    if (merchantName != null && merchantName!.isNotEmpty) {
      return merchantName!;
    }
    return '商户评价';
  }

  /// 获取格式化的人均消费
  String? get formattedPerCapita {
    if (perCapitaCost == null) return null;
    return '人均¥${perCapitaCost!.toStringAsFixed(0)}';
  }
}

/// 评价回复模型
class MerchantReviewReply {
  final int id;
  final int reviewId;
  final int? parentId;
  final int replyType; // 1-用户 2-商家 3-平台
  final int replyBy;
  final String replyName;
  final String? replyAvatar;
  final int? replyTo;
  final String? replyToName;
  final String content;
  final List<String>? images;
  final int likeCount;
  final bool official;
  final bool merchantPinned;
  final DateTime createTime;
  final List<MerchantReviewReply>? children;

  MerchantReviewReply({
    required this.id,
    required this.reviewId,
    this.parentId,
    required this.replyType,
    required this.replyBy,
    required this.replyName,
    this.replyAvatar,
    this.replyTo,
    this.replyToName,
    required this.content,
    this.images,
    this.likeCount = 0,
    this.official = false,
    this.merchantPinned = false,
    required this.createTime,
    this.children,
  });

  factory MerchantReviewReply.fromJson(Map<String, dynamic> json) {
    return MerchantReviewReply(
      id: json['id'] as int,
      reviewId: json['reviewId'] as int,
      parentId: json['parentId'] as int?,
      replyType: json['replyType'] as int,
      replyBy: json['replyBy'] as int,
      replyName: json['replyName'] as String,
      replyAvatar: json['replyAvatar'] as String?,
      replyTo: json['replyTo'] as int?,
      replyToName: json['replyToName'] as String?,
      content: json['content'] as String,
      images: json['images'] != null ? List<String>.from(json['images']) : null,
      likeCount: json['likeCount'] as int? ?? 0,
      official: json['official'] as bool? ?? false,
      merchantPinned: json['merchantPinned'] as bool? ?? false,
      createTime: DateTime.parse(json['createTime']),
      children: json['children'] != null
          ? (json['children'] as List).map((e) => MerchantReviewReply.fromJson(e)).toList()
          : null,
    );
  }

  /// 获取回复者类型标签
  String get typeLabel {
    switch (replyType) {
      case 1:
        return '用户';
      case 2:
        return '商家';
      case 3:
        return '官方';
      default:
        return '';
    }
  }

  /// 获取回复者类型颜色
  String get typeColor {
    switch (replyType) {
      case 1:
        return '#666666';
      case 2:
        return '#FF6B6B';
      case 3:
        return '#4A90E2';
      default:
        return '#666666';
    }
  }
}

/// 评价统计模型
class MerchantReviewStatistic {
  final int merchantId;
  final double overallRating;
  final double? tasteRating;
  final double? environmentRating;
  final double? serviceRating;
  final double? valueRating;
  final int totalCount;
  final int fiveStarCount;
  final int fourStarCount;
  final int threeStarCount;
  final int twoStarCount;
  final int oneStarCount;
  final int withImageCount;
  final int withVideoCount;
  final int positiveCount;
  final int neutralCount;
  final int negativeCount;
  final double positiveRate;
  final int dailyNewCount;
  final int weeklyNewCount;
  final int monthlyNewCount;
  final StarDistribution starDistribution;
  final TagStatistic tagStatistic;

  MerchantReviewStatistic({
    required this.merchantId,
    required this.overallRating,
    this.tasteRating,
    this.environmentRating,
    this.serviceRating,
    this.valueRating,
    required this.totalCount,
    required this.fiveStarCount,
    required this.fourStarCount,
    required this.threeStarCount,
    required this.twoStarCount,
    required this.oneStarCount,
    required this.withImageCount,
    required this.withVideoCount,
    required this.positiveCount,
    required this.neutralCount,
    required this.negativeCount,
    required this.positiveRate,
    required this.dailyNewCount,
    required this.weeklyNewCount,
    required this.monthlyNewCount,
    required this.starDistribution,
    required this.tagStatistic,
  });

  factory MerchantReviewStatistic.fromJson(Map<String, dynamic> json) {
    return MerchantReviewStatistic(
      merchantId: json['merchantId'] as int,
      overallRating: (json['overallRating'] as num).toDouble(),
      tasteRating: json['tasteRating'] != null ? (json['tasteRating'] as num).toDouble() : null,
      environmentRating: json['environmentRating'] != null ? (json['environmentRating'] as num).toDouble() : null,
      serviceRating: json['serviceRating'] != null ? (json['serviceRating'] as num).toDouble() : null,
      valueRating: json['valueRating'] != null ? (json['valueRating'] as num).toDouble() : null,
      totalCount: json['totalCount'] as int? ?? 0,
      fiveStarCount: json['fiveStarCount'] as int? ?? 0,
      fourStarCount: json['fourStarCount'] as int? ?? 0,
      threeStarCount: json['threeStarCount'] as int? ?? 0,
      twoStarCount: json['twoStarCount'] as int? ?? 0,
      oneStarCount: json['oneStarCount'] as int? ?? 0,
      withImageCount: json['withImageCount'] as int? ?? 0,
      withVideoCount: json['withVideoCount'] as int? ?? 0,
      positiveCount: json['positiveCount'] as int? ?? 0,
      neutralCount: json['neutralCount'] as int? ?? 0,
      negativeCount: json['negativeCount'] as int? ?? 0,
      positiveRate: (json['positiveRate'] as num).toDouble(),
      dailyNewCount: json['dailyNewCount'] as int? ?? 0,
      weeklyNewCount: json['weeklyNewCount'] as int? ?? 0,
      monthlyNewCount: json['monthlyNewCount'] as int? ?? 0,
      starDistribution: StarDistribution.fromJson(json['starDistribution'] ?? {}),
      tagStatistic: TagStatistic.fromJson(json['tagStatistic'] ?? {}),
    );
  }

  /// 获取星级文本
  String get ratingText {
    if (overallRating >= 4.5) return '极佳';
    if (overallRating >= 4.0) return '非常好';
    if (overallRating >= 3.5) return '不错';
    if (overallRating >= 3.0) return '一般';
    return '较差';
  }
}

/// 星级分布
class StarDistribution {
  final double fiveStarPercent;
  final double fourStarPercent;
  final double threeStarPercent;
  final double twoStarPercent;
  final double oneStarPercent;

  StarDistribution({
    required this.fiveStarPercent,
    required this.fourStarPercent,
    required this.threeStarPercent,
    required this.twoStarPercent,
    required this.oneStarPercent,
  });

  factory StarDistribution.fromJson(Map<String, dynamic> json) {
    return StarDistribution(
      fiveStarPercent: (json['fiveStarPercent'] as num? ?? 0).toDouble(),
      fourStarPercent: (json['fourStarPercent'] as num? ?? 0).toDouble(),
      threeStarPercent: (json['threeStarPercent'] as num? ?? 0).toDouble(),
      twoStarPercent: (json['twoStarPercent'] as num? ?? 0).toDouble(),
      oneStarPercent: (json['oneStarPercent'] as num? ?? 0).toDouble(),
    );
  }
}

/// 标签统计
class TagStatistic {
  final int tasteGoodCount;
  final int envGoodCount;
  final int serviceGoodCount;
  final int valueGoodCount;
  final int returningCount;

  TagStatistic({
    required this.tasteGoodCount,
    required this.envGoodCount,
    required this.serviceGoodCount,
    required this.valueGoodCount,
    required this.returningCount,
  });

  factory TagStatistic.fromJson(Map<String, dynamic> json) {
    return TagStatistic(
      tasteGoodCount: json['tasteGoodCount'] as int? ?? 0,
      envGoodCount: json['envGoodCount'] as int? ?? 0,
      serviceGoodCount: json['serviceGoodCount'] as int? ?? 0,
      valueGoodCount: json['valueGoodCount'] as int? ?? 0,
      returningCount: json['returningCount'] as int? ?? 0,
    );
  }
}

/// 排序类型
enum ReviewSortType {
  latest('latest', '最新'),
  highest('highest', '评分最高'),
  lowest('lowest', '评分最低'),
  liked('liked', '最多点赞'),
  recommended('recommended', '优质推荐'),
  image('image', '有图评价'),
  video('video', '视频评价');

  final String code;
  final String label;

  const ReviewSortType(this.code, this.label);

  static ReviewSortType fromCode(String? code) {
    return values.firstWhere(
      (e) => e.code == code,
      orElse: () => latest,
    );
  }
}
